sdk/servicebus/service-bus/src/core/autoLockRenewer.ts

```diff
@@ -0,0 +1,209 @@
+          // TODO: We can run into problems with clock skew between the client and the server.
+          // It would be better to calculate the duration based on the "lockDuration" property
+          // of the queue. However, we do not have the management plane of the client ready for
+          // now. Hence we rely on the lockedUntilUtc property on the message set by ServiceBus.
+          const amount = calculateRenewAfterDuration(bMessage.lockedUntilUtc!);
+          logger.verbose(
+            `${logPrefix} Sleeping for %d milliseconds while renewing the lock for message with id '${bMessage.messageId}' is: ${amount}`
+          );
+          // Setting the value of the messageId to the actual timer. This will be cleared when
+          // an attempt is made to settle the message (either by the user or by the sdk) OR
+          // when the execution of user's message handler completes.
+          this._messageRenewLockTimers.set(
+            bMessage.messageId as string,
+            setTimeout(async () => {
+              try {
+                logger.verbose(
+                  `${logPrefix} Attempting to renew the lock for message with id '${bMessage.messageId}'.`
+                );
+                bMessage.lockedUntilUtc = await this._context
+                  .getManagementClient(this._linkEntity.entityPath)
+                  .renewLock(lockToken, {
+                    associatedLinkName: this._linkEntity.name
+                  });
+                logger.verbose(
+                  `${logPrefix} Successfully renewed the lock for message with id '${bMessage.messageId}'. Starting next auto-lock-renew cycle for message.`
+                );
+                autoRenewLockTask();
+              } catch (err) {
+                logError(
+                  err,
+                  `${logPrefix} An error occured while auto renewing the message lock '${bMessage.lockToken}' for message with id '${bMessage.messageId}'`
+                );
+                onError(err);
+              }
+            }, amount)
+          );
+        } else {
+          logger.verbose(
+            `${logPrefix} Looks like the message lock renew timer has already been cleared for message with id '${bMessage.messageId}'.`
+          );
+        }
+      } else {
+        logger.verbose(
+          `${logPrefix} Current time ${new Date()} exceeds the total autolockrenew duration ${new Date(
+            totalAutoLockRenewDuration
+          )} for message with messageId '${
+            bMessage.messageId
+          }'. Hence we will stop the autoLockRenewTask.`
+        );
+
+        this.stop(bMessage);
+      }
+    };
+    // start
+    autoRenewLockTask();
+  }
+
+  private _stopAndRemoveById(messageId: string | undefined): void {
+    if (messageId == null) {
+      throw new Error("Failed to stop auto lock renewal - no message ID");
+    }
+
+    // TODO: messageId doesn't actually need to be unique. Perhaps we should use lockToken
```

Comment:
That sounds good!

Reply:
Does the service guarantee that message id will be set on the message even if the user did not set one when sending the message?

Reply:
Doesnt the lock token change after every lock renewal? If so, it cannot be used as key right?

Reply:
I will definitely check.

Reply:
Just tried this out - the only field that we modify after a lock renewal is the message.lockedUntilUtc field.

I added in some additional compile time constraints to make this more clear (specifically I ReadOnly'd the lock token field) to make this more obvious as well.